File: spring-basics/src/main/java/com/example/spring_basics/model/Headquarter.java
package com.example.spring_basics.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Table(name="headquarters")
public class Headquarter {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(length = 100, nullable = false)
    private String city;

    @Column(length = 150, nullable = true)
    private String state;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "country_id", referencedColumnName = "id", nullable = false)
    private Country country;

    @Column(length = 150, nullable = true)
    private String street;

    @Column(nullable = true)
    private Integer number;

    @Column(length = 50, nullable = true)
    private String zipCode;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "publisher_id", referencedColumnName = "id", nullable = false)
    private Publisher publisher;
}

File: angular-spring/src/app/features/headquarter/models/headquarter.ts
import { ApiResponseBase } from "../../../shared/models/response";
import { CountrySummaryItem } from "../../country/models/country";
import { PublisherSummaryItem } from "../../publisher/models/publisher";

export interface HeadquarterSummaryItem {
    id: string;
    street: string;
    city: string;
    countryName: string;
}

export interface HeadquarterResponseItem {
    id: string;
    street: string;
    city: string;
    state: string;
    number: number;
    country: CountrySummaryItem;
    publisher: PublisherSummaryItem;
}

export interface HeadquarterPageResponse extends ApiResponseBase{
    content: HeadquarterResponseItem[];
}

export interface HeadquarterCreateBody {
    city: string;
    state: string;
    street: string;
    number: number;
    zipCode: string;
    countryId: number;
    publisherId: string;
}